Radios and checkboxes without JS

robinwhittleton opened the pull request:
This PR removes the [JavaScript requirement](https://github.com/alphagov/govuk_frontend_toolkit/blob/master/javascripts/govuk/selection-buttons.js) for radio buttons and checkboxes. To do this the markup had been changed, but there are no visual changes for users (beyond IE < 9 having a focus box around just the input rather than the whole label).

#### What problem does the pull request solve?
The new radios and checkboxes still required the selection buttons JS. This is a potential point of failure that’s a hangover from the old grey-box implementation, and isn’t really needed. We originally went with it because removing it would require markup changes and would mean that it they weren’t a drop-in solution. I think that was probably a mistake though, so this corrects it.

While making the markup changes I took the opportunity to rename the component from the old `block-label` to the more descriptive `multiple-choice`. This has better developer semantics, but also has the benefit that it’ll obviously require markup changes.

This fixes https://github.com/alphagov/govuk_elements/issues/375 and https://github.com/alphagov/govuk_elements/issues/367.

#### How has this been tested?
Tested in all macOS native browsers, and in Browserstack for IE and Edge on Windows (back to 8) and a selection of iOS and Android versions. The only initial bug found was a problem in IE9/10 that was fixed. No assistive tech testing has been done yet, and I’d want to do at least some to make sure that no new issues have been introduced. The assistive tech bugs fixed during the development of the new radios and checkboxes shouldn’t have regressed as their implementation hasn’t changed.

#### What type of change is it?
- Breaking change (fix or feature that would cause existing functionality to change)

#### Has the documentation been updated?
- I have updated the documentation accordingly.

nickcolley commented:
Would this fix this issue? https://github.com/alphagov/govuk_elements/issues/354

gemmaleigh commented on public/sass/_govuk-elements.scss:
```diff
@@ -46,7 +46,7 @@
 @import "elements/panels";                        // Panels with a left grey border
 @import "elements/forms";                         // Form - wrappers, inputs, labels
-@import "elements/forms/form-block-labels";       // Chunky labels for radios and checkboxes
+@import "elements/forms/form-multiple-choice";    // Chunky labels for radios and checkboxes
```
The comment needs updating here. "Custom" radio buttons and checkboxes?

robinwhittleton commented:
No, that seems to be a generic browser rendering issue at different zoom levels (although it seems more pronounced in Chrome than other browsers). This doesn’t change the CSS rules that render the inputs (apart from fixing that IE bug).

nickcolley commented:
@robinwhittleton it seems as though, if we change the markup to add specific elements instead of pseudo elements we'd be able to fix that issue?

robinwhittleton commented:
Honestly I haven’t tried it, but I’m not sure it would make any difference. I’ll give it a shot though.

robinwhittleton commented:
I’ve tried out a few different ways of fixing https://github.com/alphagov/govuk_elements/issues/354 as part of this issue, but nothing is looking particularly promising. I’m loathe to complicate the markup here, especially as adding empty spans can cause issues for JAWS-on-IE and requires extra fragile workarounds. In the componentised future this wouldn’t be a problem, but where we’re relying on people cutting and pasting HTML I’m worried about the potential for accessibility problems in a small visual fix for a non-standard setup. It’s worth pointing out that the visual problem only manifests itself when the page is zoomed _out_, an unusual thing to do compared to zooming in.

tl;dr it’s worth leaving the issue open, but I don’t think we need to address it in this PR.

robinwhittleton commented:
Blocked by https://github.com/alphagov/govuk_elements/pull/394/files

nickcolley commented:
@robinwhittleton do you have a link to the the empty spans problem?

robinwhittleton commented:
It was found and fixed on an internal repo, but the gist is that with the markup:

```html
<label>
  <input type="radio" />
  <span><span></span></span>
  Example
</label>
```

The label text wouldn’t be read by JAWS-on-IE (JAWS-on-Firefox was fine). Some text needed to be added to the middle of the elements (we settled on `<span><span>&nbsp;</span></span>`) to get it read out. The old Verify custom radio builder shows this: https://github.com/alphagov/verify-frontend/blob/4f5b1592bc3e861dfd3f2294ddf9f816ceca01d3/lib/verify_form_builder.rb

nickcolley commented:
Very interesting, I wonder if that could be solved with `aria-hidden`?

robinwhittleton commented:
Potentially? In other news, I’ve rebased this and updated on top of [Gemma’s work to convert the joint text box + checkbox over to the new style](https://github.com/alphagov/govuk_elements/pull/394) so this is now unblocked for merging.

robinwhittleton commented:
…assuming the accessibility team are happy. @cfq / @accessiblewebuk / @selfthinker ?

selfthinker commented:
Should we give it another full testing round? Or should we only have a quick look? Or do you want us to test anything specific?

robinwhittleton commented:
Well, the whole markup has changed, but the CSS remains mostly the same. I guess we probably want to retest it in everything based on that?

selfthinker commented:
This is now at the top of our backlog, Richard will probably test these soon.
Do you have a URL for him to test?

robinwhittleton commented:
I’ve pushed the branch to http://govuk-elements-wip.herokuapp.com/form-elements/#form-radio-buttons . That’s a temporary site, but it’s only really @gemmaleigh and I who use it so we’ll let you know if we need to repurpose it.

accessiblewebuk commented:
I have completed a full round of accessibility testing (including WCAG 2.0 A and AA, JAWS14, JAWS17, NVDA 2017, Dragon 12.5 and 13, Read and Write Gold and Zoomtext) with the exception of VoiceOver on Mac which I can't do. No issues found.

robinwhittleton commented:
I integrated this into [a branch on my copy of Verify’s codebase](https://github.com/alphagov/verify-frontend/compare/master...robinwhittleton:new-new-custom-radios) (with a custom built gem). There wasn’t too much work in the end, it was mostly down to updating the markup in the tests.

robinwhittleton commented:
When this is merged and released we’ll want to deprecate selection-buttons.js in govuk_frontend_toolkit. See https://github.com/alphagov/govuk_frontend_toolkit/pull/389

gemmaleigh commented:
I'm happy to merge, this has been reviewed by the accessibility team - thanks @accessiblewebuk and @robinwhittleton!
